File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    
    private static int[] dR = new int[]{-1, 1, 0, 0};
    private static int[] dC = new int[]{0, 0, -1, 1};
    
    public int solution(int[][] maps) {
        int answer = 0;
        answer = getAnswer(maps);
        return answer;
    }
    
    private int getAnswer(int[][] maps) {
        boolean[][] visited = new boolean[maps.length][maps[0].length];
        Queue<int[]> queue = new ArrayDeque<>();
        queue.offer(new int[]{0,0,1});
        visited[0][0] = true;
        
        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            int cR = cur[0];
            int cC = cur[1];
            int cnt = cur[2];
            
            if (cR == maps.length - 1 && cC == maps[0].length - 1) {
                return cnt;
            }
            
            for (int i = 0; i < 4; i++) {
                int nR = cR + dR[i];
                int nC = cC + dC[i];
                
                if (!isIn(nR, nC, maps.length, maps[0].length)) {
                    continue;
                }
                if (visited[nR][nC] || maps[nR][nC] == 0) {
                    continue;
                }
                visited[nR][nC] = true;
                queue.offer(new int[]{nR, nC, cnt + 1});
            }   
        }        
        return -1;
    }
    
    private boolean isIn(int r, int c, int R, int C) {
        return r >= 0 && r < R && c >= 0 && c < C;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    global answer, dr, dc, map_view, visited
    answer = -1
    dr = [1, -1, 0, 0]
    dc = [0, 0, -1, 1]
    map_view = maps
    visited = []
    for i in range(len(map_view)):
        row = []
        for j in range(len(map_view[0])):
            row.append(False)
        visited.append(row)
    visited[0][0] = True
    result = bfs()
    if result is not None:
        answer = result
    return answer
    

def bfs():
    queue = deque([])
    queue.append([0,0,1])
    while queue:
        cur = queue.popleft()
        cr = cur[0]
        cc = cur[1]
        if cr == len(map_view) - 1 and cc == len(map_view[0]) - 1:
            return cur[2]
        
        for i in range(4):
            nr = cr + dr[i]
            nc = cc + dc[i]
            if not is_in(nr, nc):
                continue
            if visited[nr][nc] or map_view[nr][nc] == 0:
                continue
            visited[nr][nc] = True
            queue.append([nr, nc, cur[2] + 1])
                

def is_in(r, c) -> bool:
    return r >= 0 and r < len(map_view) and c >= 0 and c < len(map_view[0])